src/enumspecindex.ts

```diff
@@ -0,0 +1,134 @@
+  autoCount?: EnumSpec<AggregateOp>;
+
+  /** Enum spec for bin enumeration. */
+  bin?: EnumSpec<boolean>;
+
+  /** Enum spec for bin.maxbins enumeration */
+  maxbin?: EnumSpec<number>;
+
+  /** Enum spec for scale enumeration. */
+  scale?: EnumSpec<boolean>;
+
+  /** Enum spec for scale.bandSize enumeration */
+  scaleBandSize?: EnumSpec<number>;
+
+  /** Enum spec for scale.clamp enumeration */
+  scaleClamp?: EnumSpec<boolean>;
+
+  /** Enum spec for scale.domain enumeration */
+  scaleDomain?: EnumSpec<string | string[] | number[]>;
+
+  /** Enum spec for scale.exponent enumeration */
+  scaleExponent?: EnumSpec<number>;
+
+  /** Enum spec for scale.nice enumeration */
+  scaleNice?: EnumSpec<boolean>;
+
+  /** Enum spec for scale.range enumeration */
+  scaleRange?: EnumSpec<string | string[] | number[]>;
+
+  /** Enum spec for scale.round enumeration */
+  scaleRound?: EnumSpec<boolean>;
+
+  /** Enum spec for scale.type enumeration */
+  scaleType?: EnumSpec<ScaleType>;
+
+  /** Enum spec for scale.useRawDomain enumeration */
+  scaleUseRawDomain?: EnumSpec<boolean>;
+
+  /** Enum spec for scale.zero enumeration */
+  scaleZero?: EnumSpec<boolean>;
+
+  /** Enum spec for timeUnit enumeration. */
+  timeUnit?: EnumSpec<TimeUnit>;
+
+  /** Enum spec for field enumeration. */
+  field?: EnumSpec<string>;
+
+  /** Enum spec for type enumeration. */
+  type?: EnumSpec<Type>;
+}
+
+export interface EncodingsEnumSpecIndex {
+  [index: number]: EncodingEnumSpecIndex;
+}
+
+export class EnumSpecIndex {
+  private _mark: EnumSpec<Mark>;
+  // TODO: transform
+
+  /**
+   * Dictionary mapping encoding index to an encoding enum spec index.
+   */
+
+  private _encodings: EncodingsEnumSpecIndex;
+  private _encodingIndicesByProperty: Dict<number[]>;
+
+  constructor(mark: EnumSpec<Mark>, encodings: EncodingsEnumSpecIndex, encodingIndicesByProperty: Dict<number[]>) {
+    this._mark = mark;
+    this._encodings = encodings;
+    this._encodingIndicesByProperty = encodingIndicesByProperty;
+  }
+
+  public set(index: number, prop: Property, enumSpec: EnumSpec<any>) {
+    const encodingsIndex = this._encodings;
+
+    // Init encoding index and set prop
+    const encIndex = encodingsIndex[index] = encodingsIndex[index] || {};
+    encIndex[prop] = enumSpec;
+
+    // Initialize indicesByProperty[prop] and add index
+    const encodingIndicesByProperty = this._encodingIndicesByProperty;
+    (encodingIndicesByProperty[prop] = encodingIndicesByProperty[prop] || []).push(index);
+  }
+
+  public hasProperty(prop: Property) {
+    if (isEncodingProperty(prop)) {
+      return !!this.encodingIndicesByProperty[prop];
+    } if (prop === Property.MARK) {
+      return !!this.mark;
+    }
+    /* istanbul ignore next */
+    throw new Error('Unimplemented for property ' + prop);
+  }
+
+  public isEmpty() {
+    return !this.mark && keys(this.encodingIndicesByProperty).length === 0;
+  }
+
+  public set mark(mark) {
```

Comment:
Since we'll have `setEncodingProperty`,  maybe let's make this a normal method (`setMark()`) for consistency.  

Also make both of these methods support chaining just like D3 (return `this`) and use it for testing.